### PR TITLE
Created the terraform for the API

### PR DESCRIPTION
**Pull request by CharliedeanCS**
Created a new ECR repo in Terraform code for the API.
- Creates a repo for the API so a docker image of the API can be used.
Created a Task definition in Terraform code for the API.
- Creates a task definition to run the API Image.

Created a security group for the API.
- Allows access on port 5000
Created a ECS Service for the API.
- Runs the API Task def

Changed the Terraform README.

### CODE REVIEW

File: api/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM python:latest
+
+WORKDIR /api
+
+COPY requirements.txt .
+RUN pip install -r requirements.txt
+
+COPY . /api
```
Comment (dawiddawidowski) [inline api/Dockerfile:8]: Nice ordering of the copied files

File: api/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+EXPOSE 5000
+
+ENV FLASK_APP=api.py
+
+CMD ["flask", "run", "--host=0.0.0.0"]
```
Comment (dawiddawidowski) [inline api/Dockerfile:14]: Very good Dockerfile

File: terraform/README.md [hunk 3]
```diff
@@ -8,15 +8,15 @@ This folder should contain all code and resources required to handle the infrast
...
-- `database_name`
+- `DATABASE_USERNAME`
+- `DATABASE_PASSWORD`
+- `AWS_ACCESS_KEY_ID_`
+- `AWS_SECRET_ACCESS_KEY_`
+- `DATABASE_USERNAME`
+- `DATABASE_PASSWORD`
+- `DATABASE_IP`
+- `DATABASE_PORT`
+- `DATABASE_NAME`
```
Comment (dawiddawidowski) [inline terraform/README.md:19]: Good documentation to ensure correct letter case

File: terraform/README.md [hunk 4]
```diff
@@ -30,9 +30,25 @@ Remove the terraform with `terraform destroy`
...
+     - `Pipeline Task Definition`
+      - A task definition that runs the pipeline image.
+     - `Pipeline ECS Service`
+      - A ECS Service that runs the pipeline Task Definition.
+     - `ECR API`
+      - An ECR to store the API container.
+     - `API Task Definition`
+      - A task definition that runs the API image.
+     - `API Security Group`
+      - Security Group setup for the API to allow access on port 5000.
```
Comment (dawiddawidowski) [inline terraform/README.md:52]: Super clear and concise description of each file.

File: terraform/main.tf [hunk 5]
```diff
@@ -464,4 +469,90 @@ resource "aws_ecs_service" "pipeline_service" {
...
+  execution_role_arn       = "${data.aws_iam_role.ecs_task_execution_role.arn}"
+  container_definitions    = <<TASK_DEFINITION
+[
+  {
+    "environment": [
+      {"name": "DATABASE_IP", "value": "${var.DATABASE_IP}"},
+      {"name": "DATABASE_NAME", "value": "${var.DATABASE_NAME}"},
+      {"name": "DATABASE_PASSWORD", "value": "${var.DATABASE_PASSWORD}"},
+      {"name": "DATABASE_PORT", "value": "${var.DATABASE_PORT}"},
+      {"name": "DATABASE_USERNAME", "value": "${var.DATABASE_USERNAME}"}
```
Comment (dawiddawidowski) [inline terraform/main.tf:496]: I like this concise method of adding the environmental variables.

File: terraform/README.md [hunk 6]
```diff
@@ -30,9 +30,25 @@ Remove the terraform with `terraform destroy`
...
+     - `API ECS Service`
+      - A ECS Service that runs the API Task Definition.
```
Comment (lcawood) [inline terraform/README.md:54]: This is solid documentation

File: terraform/main.tf [hunk 7]
```diff
@@ -464,4 +469,90 @@ resource "aws_ecs_service" "pipeline_service" {
...
+    to_port          = 5000
+    protocol         = "tcp"
+    cidr_blocks      = ["0.0.0.0/0"] 
+  }
+  egress {
+    from_port        = 0
+    to_port          = 0
+    protocol         = "-1"
+    cidr_blocks      = ["0.0.0.0/0"]
+    ipv6_cidr_blocks = ["::/0"]
```
Comment (dawiddawidowski) [inline terraform/main.tf:536]: Nice inbound and outbound rule setup

File: terraform/main.tf [hunk 8]
```diff
@@ -464,4 +469,90 @@ resource "aws_ecs_service" "pipeline_service" {
...
+  cluster         = "c9-ecs-cluster"
+  task_definition = aws_ecs_task_definition.c9_deloton_api_task_def_t.arn
+  desired_count   = 1
+  launch_type     = "FARGATE"
+  force_new_deployment = true 
+  depends_on = [aws_ecs_task_definition.c9_deloton_api_task_def_t]
+
+network_configuration {
+    security_groups = [aws_security_group.c9_velo_securitygroup_api.id]
+    subnets         = ["subnet-0d0b16e76e68cf51b","subnet-081c7c419697dec52","subnet-02a00c7be52b00368"]
```
Comment (dawiddawidowski) [inline terraform/main.tf:555]: Good use of only public subnets to allow public access to the API

**Review by dawiddawidowski — APPROVED** [review]

File: terraform/main.tf [hunk 9]
```diff
@@ -464,4 +469,90 @@ resource "aws_ecs_service" "pipeline_service" {
...
+    to_port          = 0
+    protocol         = "-1"
+    cidr_blocks      = ["0.0.0.0/0"]
+    ipv6_cidr_blocks = ["::/0"]
+  }
+
+  tags = {
+    Name = "c9_velo_securitygroup_api"
+  }
+}
```
Comment (lcawood) [inline terraform/main.tf:542]: This all looks good to me

**Pull request merged**